### DISCUSSION TITLE
How to build for all architectures?

### DISCUSSION THREAD

**Discussion by humanfriend22** [tauri-apps/tauri#4528] [Q&A] [answered]
I need to build for:

### Mac:
- M1 Silicon
- Intel

### Windows:
- Intel x64
- amd x64

### Linux:
- armv6
- armv7
- Intel x64
- amd x64

I also need all this to automated by Github Actions but I am not familiar with the process. Thank you in advance.

**Comment by Uzaaft** [comment]
The following guide should help you:
https://tauri.app/v1/guides/building/cross-platform/

**Reply by humanfriend22** [reply]
Yeah, I could use Github Actions but I found compiling for Raspberry Pi a little difficult so I'm probably going to use [cross](https://github.com/cross-rs/cross) mentioned [here](https://github.com/tauri-apps/tauri/discussions/4528#discussioncomment-3100794). Thank you though.

**Comment by simonhyll** [comment, marked as answer]
There's channels you can look through on the Discord for various architectures. E.g. arm is something people have worked with quite a bunch. There's a lot of stuff to consider when doing cross compiling. I'm currently pursuing using this project: <https://github.com/cross-rs/cross>

**Reply by Sytten** [reply]
Would you mind detailing how you used cross exactly?

**Reply by simonhyll** [reply]
> Would you mind detailing how you used cross exactly?

First I set it up locally according to their instructions. Then I started working on an image that has all the Tauri dependencies on it. I tried ot build it on Github first but for very techy reasons Docker can build differently on Github than on my own computer, and I had fewer issues on my own computer, so I built an image and pushed it to Docker Hub that Github could then use (saves a lot of Github Action time anyway). The problem I was having when I last looked at it was with dependencies, weird issues, issues where when I fix one issue in one line, another line creates 5 more issues, and when I fix those 5 issues, one of the old issues becomes an issue again. Very weird stuff. It's still the best bet for cross compiling on Github Actions I feel, but I've decided to do other things for a while until I set up my raspberry pi again and can have a nicer setup for debugging and researching the dependency issues I'm having.

**Comment by joaoantoniocardoso** [comment]
I'm building (`.deb`) tauri app for a Raspberry Pi 3 (armv7), managed to do as follows:

```Dockerfile
FROM debian:bullseye-slim
ARG DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt-get install --assume-yes --no-install-recommends \
        gnupg2 \
        ca-certificates \
        g++-arm-linux-gnueabihf \
        libc6-dev-armhf-cross \
        libclang-3.9-dev \
        build-essential \
        pkg-config \
        curl \
        wget \
        file

RUN dpkg --add-architecture armhf && \
    apt-get update && apt-get install --assume-yes --no-install-recommends \
        libwebkit2gtk-4.0-dev:armhf \
        libssl-dev:armhf \
        libgtk-3-dev:armhf \
        libayatana-appindicator3-dev:armhf \
        librsvg2-dev:armhf \
        libatk1.0-dev:armhf \
        libgdk-pixbuf-2.0-dev:armhf \
        libcairo2-dev:armhf \
        libpango1.0-dev:armhf \
        libgtk-3-dev:armhf \
        libsoup2.4-dev:armhf

# Install NodeJs and Yarn
RUN curl -fsSL https://deb.nodesource.com/setup_19.x | bash - &&\
    curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add &&\
    echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list &&\
    apt-get update && apt-get install --assume-yes --no-install-recommends \
        nodejs \
        yarn

# Install Rust
RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --target armv7-unknown-linux-gnueabihf

# Set envs
ENV CARGO_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_LINKER=arm-linux-gnueabihf-gcc \
    CC_armv7_unknown_linux_gnueabihf=arm-linux-gnueabihf-gcc \
    CXX_armv7_unknown_linux_gnueabihf=arm-linux-gnueabihf-g++ \
    BINDGEN_EXTRA_CLANG_ARGS_armv7_unknown_linux_gnueabihf="--sysroot=/usr/arm-linux-gnueabihf" \
    PKG_CONFIG_PATH="/usr/lib/arm-linux-gnueabihf/pkgconfig/:${PKG_CONFIG_PATH}" \
    PKG_CONFIG_ALLOW_CROSS=1 \
    PATH="/root/.cargo/bin:$PATH"
WORKDIR /app
```

I built it like:
```Bash
docker build \
    -t $USER/tauri-cross:armv7-unknown-linux-gnueabihf \
    --output type=registry \
    --progress=plain \
    docker/armv7-unknown-linux-gnueabihf/
```

and then I run with `yarn tauri build` command as follows:
```Bash
docker run --rm -it -v $PWD:/app \
    $USER/tauri-cross:armv7-unknown-linux-gnueabihf bash \
    -c "yarn tauri build --target armv7-unknown-linux-gnueabihf"
```


Hope it helps

**Reply by pkh1979** [reply]
This is great! however do you have aarch64 / arm64 version of these script?

**Reply by joaoantoniocardoso** [reply]
I do not, but I think it is just a matter of replacing the `armhf` and `armv7` strings to the desired architectures.

**Reply by PMLS3** [reply]
@pkh1979 did you come right with a aarch64/arm64 version? If you did any possibilities of sharing?

**Reply by FabianLars** [reply]
It is indeed as simple as replacing the strings. We also have a guide on this (no docker though) [here](https://tauri.app/v1/guides/building/linux#cross-compiling-tauri-applications-for-arm-based-devices) if that helps.
